### PR TITLE
fix doc for functional.dropout*

### LINKED ISSUE #4177: Better functional.dropout{,2d,3d} docs
http://pytorch.org/docs/master/nn.html#torch.nn.functional.dropout

### PR DESCRIPTION
**Pull request by weiyangfb**
- fixes #4177

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -591,7 +591,19 @@ def adaptive_avg_pool3d(input, output_size):
 
 
 # Activation functions
-def dropout(input, p=0.5, training=False, inplace=False):
+def dropout(input, p=0.5, training=True, inplace=False):
```
Comment (SsnL) [inline torch/nn/functional.py]: please do not change the default arg.

**Review by ssnl — CHANGES_REQUESTED** [review]

File: torch/nn/functional.py (hunk 1 above)
Comment (ezyang) [reply torch/nn/functional.py]: I agree that it is kind of goofy. If you don't pass a `train` parameter, functional dropout is always a no-op and there's no point in calling it. (I don't know what the BC breaking implications of this are though.)

Comment (weiyangfb) [reply torch/nn/functional.py]: That's true, I will revert the default arg and make a note at the API doc. Still wonder when will `correcting things` have higher priority than `not BC breaking`..

Comment (ezyang) [reply torch/nn/functional.py]: I think one could reasonably make an argument that changing the default here will not break code that was already "broken" (dropout silently ignored.)

Comment (weiyangfb) [reply torch/nn/functional.py]: yeah, I agree. Although it might still break someone's code if they understand the default behavior.. I can go with either way. Thoughts? cc @SsnL

Comment (SsnL) [reply torch/nn/functional.py]: I'm fine with the change. I personally welcome it as a user as I discovered this difference in an unpleasant way. I don't think that there is much code assuming F.dropout has default `training=False` (because it would be noop w/o specification). But I'm kinda sad we can't give a warning somehow.

Comment (weiyangfb) [reply torch/nn/functional.py]: you mean I should put a warning there after changing the default arg?

Comment (SsnL) [reply torch/nn/functional.py]: alright let's just change it.

you can't make a warning because it is impossible to differentiate which users want from the calling args.

File: torch/nn/modules/dropout.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def forward(self, input):
 
 
 class Dropout2d(_DropoutNd):
-    r"""Randomly zeroes whole channels of the input tensor.
+    r"""Randomly zeroes whole channels (a channel is a (N, C) pair) of the input tensor.
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: This is a bit confusing. A channel is in fact a 2d slice of the last two dimensions.

File: torch/nn/functional.py [hunk 3]
```diff
@@ -613,15 +625,35 @@ def alpha_dropout(input, p=0.5, training=False, inplace=False):
...
-def dropout3d(input, p=0.5, training=False, inplace=False):
+def dropout3d(input, p=0.5, training=True, inplace=False):
+    r"""
+    Randomly zeroes whole channels (a channel is a 3D slice of dimensions D, H, W) of the input tensor.
+
+    See :class:`~torch.nn.Dropout3d` for details.
+
+    Args:
+        p: probability of an element to be zeroed. Default: 0.5
+        training: apply dropout if is True. Defualt: True
```
Comment (li-roy) [inline torch/nn/functional.py]: nit: format these booleans in all the doc strings like you did for inplace

File: torch/nn/functional.py [hunk 4]
```diff
@@ -613,15 +625,35 @@ def alpha_dropout(input, p=0.5, training=False, inplace=False):
...
         raise ValueError("dropout probability has to be between 0 and 1, "
                          "but got {}".format(p))
     f = _VF.feature_dropout_ if inplace else _VF.feature_dropout
     return f(input, p, training)
 
 
-def dropout3d(input, p=0.5, training=False, inplace=False):
+def dropout3d(input, p=0.5, training=True, inplace=False):
+    r"""
+    Randomly zeroes whole channels (a channel is a 3D slice of dimensions D, H, W) of the input tensor.
```
Comment (li-roy) [inline torch/nn/functional.py]: We shouldn't refer to D H W here, it isn't defined in this context and only makes sense in the nn doc.

**Review by li-roy — COMMENTED** [review]
looks good, couple of nits. only commented on dropout3d but applies to all of them

File: torch/nn/functional.py (hunk 4 above)
Comment (SsnL) [reply torch/nn/functional.py]: yeah. although the original doc indeed wasn't very clear on what is "channel", and whether each sample in a batch can have different channels zeroed out.

Comment (weiyangfb) [reply torch/nn/functional.py]: I agree. I will remove `D H W` here. And maybe we should define "channel" at `conv` ops. By "different channels", do you mean "zeros a channel independently"?

File: torch/nn/functional.py (hunk 3 above)
Comment (weiyangfb) [reply torch/nn/functional.py]: right, I will change it

File: torch/nn/functional.py (hunk 4 above)
Comment (SsnL) [reply torch/nn/functional.py]: yep. 1st sample can have its 1st channel zeroed out, while the 2nd sample doesn't.

File: torch/nn/modules/dropout.py [hunk 5]
```diff
@@ -54,8 +55,10 @@ def forward(self, input):
 
 
 class Dropout2d(_DropoutNd):
-    r"""Randomly zeroes whole channels of the input tensor.
-    The channels to zero-out are randomized on every forward call.
+    r"""Randomly zero-out entire channels (a channel a 2D feature map of
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: nit: missing `is`

File: torch/nn/modules/dropout.py [hunk 6]
```diff
@@ -20,7 +20,8 @@ def extra_repr(self):
 class Dropout(_DropoutNd):
     r"""During training, randomly zeroes some of the elements of the input
     tensor with probability :attr:`p` using samples from a Bernoulli
-    distribution. The elements to zero are randomized on every forward call.
+    distribution. ach channel will be zero-out indipendently on every forward
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: nit: ach?

**Comment by weiyangfb** [comment]
@SsnL is this good to go?

File: torch/nn/modules/dropout.py [hunk 7]
```diff
@@ -93,8 +96,10 @@ def forward(self, input):
 
 
 class Dropout3d(_DropoutNd):
-    r"""Randomly zeroes whole channels of the input tensor.
-    The channels to zero are randomized on every forward call.
+    r"""Randomly zero-out entire channels (a channel is a 3D feature map of
+    dimensions D, H, W) of the input tensor. Each channel will be zero-out
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: be zero*ed* out. also i would drop the `-`.

File: torch/nn/modules/dropout.py [hunk 8]
```diff
@@ -54,8 +55,10 @@ def forward(self, input):
 
 
 class Dropout2d(_DropoutNd):
-    r"""Randomly zeroes whole channels of the input tensor.
-    The channels to zero-out are randomized on every forward call.
+    r"""Randomly zero-out entire channels (a channel is a 2D feature map of
+    dimensions H, W) of the input tensor. Each channel will be zero-out
+    indipendently on every forward call. with probability :attr:`p` using
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: independently

File: torch/nn/modules/dropout.py [hunk 9]
```diff
@@ -54,8 +55,10 @@ def forward(self, input):
 
 
 class Dropout2d(_DropoutNd):
-    r"""Randomly zeroes whole channels of the input tensor.
-    The channels to zero-out are randomized on every forward call.
+    r"""Randomly zero-out entire channels (a channel is a 2D feature map of
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: I feel that the explanation of channel here isn't very clear still. People who don't know the concept will still be confused how this works. Maybe say something like `` (e.g., the `j`-th channel of the `i`-th sample in the batched input is the 2D tensor `input[i, j]`) ``

File: torch/nn/modules/dropout.py [hunk 10]
```diff
@@ -54,8 +55,11 @@ def forward(self, input):
 
 
 class Dropout2d(_DropoutNd):
-    r"""Randomly zeroes whole channels of the input tensor.
-    The channels to zero-out are randomized on every forward call.
+    r"""Randomly zero out entire channels (a channel is a 2D feature map,
+    e.g., the :math:`j`-th channel of the :math:`i`-th sample in the
+    batched input is a 2D tensor :math:`input[i, j]`) of the input tensor).
```
Comment (SsnL) [inline torch/nn/modules/dropout.py]: nit: Could you add `\text` wrapper on `input` please?

**Review by ssnl — APPROVED** [review]
LGTM! Thanks!
